Blish HUD/GameServices/ArcDps/SocketListener/SocketListener.cs

```diff
@@ -35,142 +46,216 @@ public sealed class SocketListener {
             try {
-                if (listenSocket.Connected) {
-                    listenSocket.Shutdown(SocketShutdown.Receive);
+                if (socket.Connected) {
+                    socket.Shutdown(SocketShutdown.Receive);
+                    socket.Close();
+
+                    Logger.Debug("Disconnected.");
+                } else {
+                    Logger.Warn("Socket already disconnected.");
                 }
 
-                listenSocket.Close();
+                this.Running = false;
             } catch (Exception reason) {
-                Logger.Warn(reason, "An unexpected error occurred when disconnecting the Arcdps-BHUD bridge.");
+                Logger.Error(reason, "Failed to disconnect socket:");
             }
         }
 
-        private void OnIoCompleted(object sender, SocketAsyncEventArgs e) {
-            switch (e.LastOperation) {
-                case SocketAsyncOperation.Receive:
-                    ProcessReceive(e);
-                    break;
-                case SocketAsyncOperation.Connect:
-                    ProcessConnect(e);
-                    break;
-            }
-        }
+        private void StartConnect(Socket client, EndPoint endPoint) {
+            try {
+                _ = client.BeginConnect(endPoint, this.ConnectCallback, client);
+                Logger.Debug("Connected.");
+            } catch (Exception ex) {
+                Logger.Error(ex, "Failed to connect socket.");
 
-        private void ProcessConnect(SocketAsyncEventArgs e) {
-            if (e.SocketError != SocketError.Success) {
-                this.OnSocketError?.Invoke(this, e.SocketError);
-                return;
+                this.Stop();
             }
+        }
 
+        private void ConnectCallback(IAsyncResult ar) {
             try {
-                e.UserToken = new AsyncUserToken();
+                // Retrieve the socket from the state object.
+                Socket socket = (Socket)ar.AsyncState;
 
-                // The next line returns true when the operation is pending; false when it completed without delay
-                if (!e.AcceptSocket.ReceiveAsync(e)) {
-                    ProcessReceive(e);
-                }
-            } catch {
-                // ignored
+                // Complete the connection.
+                socket.EndConnect(ar);
+
+                Logger.Debug("Socket connected to {0}",
+                    socket.RemoteEndPoint.ToString());
+
+                this.Running = true;
+
+                this.StartReceive(socket);
+            } catch (Exception ex) {
+                Logger.Error(ex, "Failed to connect socket:");
+
+                this.Stop();
             }
         }
 
-        private void ProcessReceive(SocketAsyncEventArgs e) {
-            var token = (AsyncUserToken)e.UserToken;
+        private void StartReceive(Socket socket, SocketState state = null, int? retries = null) {
+            retries ??= _retryReceiveCount;
```

Comment:
Optional: If you change the _retryReceiveCount to be const, you could inline it directly in the function signature

Reply:
No idea why I didn't do that but I have changed that!